File: server/assets/solutions/ohtu2020/part4/IntJoukkoSovellus/src/main/java/ohtu/intjoukkosovellus/IntJoukko.java

package ohtu.intjoukkosovellus;

public class IntJoukko {

    public final static int KAPASITEETTI = 5, // aloitustalukon koko
                            OLETUSKASVATUS = 5;  // luotava uusi taulukko on 
    // näin paljon isompi kuin vanha
    private int kasvatuskoko;     // Uusi taulukko on tämän verran vanhaa suurempi.
    private int[] luvut;      // Joukon luvut säilytetään taulukon alkupäässä. 
    private int alkioidenLkm;    // Tyhjässä joukossa alkioiden_määrä on nolla. 

    public IntJoukko() {
        this(KAPASITEETTI, OLETUSKASVATUS);
    }

    public IntJoukko(int kapasiteetti) {
        this(kapasiteetti, OLETUSKASVATUS);
    }
    
    
    public IntJoukko(int kapasiteetti, int kasvatuskoko) {
        if (kapasiteetti < 0 || kasvatuskoko < 0) {
            throw new IndexOutOfBoundsException("Kapasiteetin ja kasvatuskoon pitää olla positiivisia");
        }

        luvut = new int[kapasiteetti];
        alkioidenLkm = 0;
        this.kasvatuskoko = kasvatuskoko;

    }

    public boolean lisaa(int luku) {
        
        if (kuuluu(luku)) {
            return false;
        }
        
        luvut[alkioidenLkm] = luku;
        alkioidenLkm++;
        
        if (alkioidenLkm % luvut.length == 0) {
            int[] uusi = new int[alkioidenLkm + kasvatuskoko];
            kopioiTaulukko(luvut, uusi);
            luvut = uusi;
        }

        return true;
    }

    public boolean kuuluu(int luku) {
        return paikassa(luku)>-1;
    }

    private int paikassa(int luku) {
        for (int i = 0; i < alkioidenLkm; i++) {
            if (luku == luvut[i]) {
                return i;
            }
        }
        
        return -1;
    }
    
    public boolean poista(int luku) {
        int kohta = paikassa(luku);
        if (kohta == -1) {
            return false;
        }

        for (int j = kohta; j < alkioidenLkm - 1; j++) {
            luvut[j] = luvut[j + 1];
        }
        
        alkioidenLkm--;
        return true;    
    }

    private void kopioiTaulukko(int[] vanha, int[] uusi) {
        for (int i = 0; i < vanha.length; i++) {
            uusi[i] = vanha[i];
        }

    }

    public int mahtavuus() {
        return alkioidenLkm;
    }

    @Override
    public String toString() {
        if (alkioidenLkm == 0) {
            return "{}";
        }
        
        String tuotos = "";
        for (int i = 0; i < alkioidenLkm - 1; i++) {
            tuotos += luvut[i];
            tuotos += ", ";
        }
        tuotos += luvut[alkioidenLkm - 1];

        return "{" + tuotos + "}";

    }

    public int[] toIntArray() {
        int[] taulu = new int[alkioidenLkm];
        for (int i = 0; i < taulu.length; i++) {
            taulu[i] = luvut[i];
        }
        return taulu;
    }
   

    public static IntJoukko yhdiste(IntJoukko a, IntJoukko b) {
        IntJoukko tulos = new IntJoukko();
        for (int alkio : a.toIntArray()) {
            tulos.lisaa(alkio);
        }
        for (int alkio : b.toIntArray()) {
            tulos.lisaa(alkio);
        }  

        return tulos;
    }

    public static IntJoukko leikkaus(IntJoukko a, IntJoukko b) {
        IntJoukko tulos = new IntJoukko();
        for (int alkio : a.toIntArray()) {
            if (b.kuuluu(alkio)) {
                tulos.lisaa(alkio);
            }     
        }

        return tulos;
    }
    
    public static IntJoukko erotus (IntJoukko a, IntJoukko b) {
        IntJoukko tulos = new IntJoukko();
        for (int alkio : a.toIntArray()) {
            if (!b.kuuluu(alkio)) {
                tulos.lisaa(alkio);
            }     
        }
        
        return tulos;
    }
        
}

File: server/assets/solutions/ohtu-avoin-2022/part4/python/int-joukko/src/int_joukko.py
KAPASITEETTI = 5
OLETUSKASVATUS = 5


class IntJoukko:
    def __init__(self, kapasiteetti=KAPASITEETTI, kasvatuskoko=OLETUSKASVATUS):
        if not isinstance(kapasiteetti, int) or kapasiteetti < 0:
            raise Exception("Väärä kapasiteetti") 
        if not isinstance(kapasiteetti, int) or kapasiteetti < 0:
            raise Exception("Väärä kapasiteetti")         
          
        self.kapasiteetti = kapasiteetti          
        self.kasvatuskoko = kasvatuskoko

        self.luvut = [0] * self.kapasiteetti
        self.alkioiden_lkm = 0

    def kuuluu(self, luku):
        return self.paikka(luku) > -1

    def paikka(self, luku):
        for i in range(self.alkioiden_lkm):
            if luku == self.luvut[i]:
                return i 

        return -1     

    def lisaa(self, luku):
        if self.kuuluu(luku):
          return False

        self.luvut[self.alkioiden_lkm] = luku
        self.alkioiden_lkm = self.alkioiden_lkm + 1

        if self.alkioiden_lkm == len(self.luvut):
            uusi = [0] * (self.alkioiden_lkm + self.kasvatuskoko)
            self.kopioi_taulukko(self.luvut, uusi)
            self.luvut = uusi

        return True   

    def poista(self, luku):
        kohta = self.paikka(luku) 
        if kohta == -1:
            return False

        for j in range(kohta, self.alkioiden_lkm - 1):
            self.luvut[j] = self.luvut[j + 1]

        self.alkioiden_lkm = self.alkioiden_lkm - 1
        return True

    def kopioi_taulukko(self, a, b):
        for i in range(0, len(a)):
            b[i] = a[i]

    def mahtavuus(self):
        return self.alkioiden_lkm

    def to_int_list(self):
        taulu = [0] * self.alkioiden_lkm

        for i in range(0, len(taulu)):
            taulu[i] = self.luvut[i]

        return taulu

    @staticmethod
    def yhdiste(a, b):
        tulos = IntJoukko()

        for luku in a.to_int_list():
          tulos.lisaa(luku)

        for luku in b.to_int_list():
          tulos.lisaa(luku)

        return tulos

    @staticmethod
    def leikkaus(a, b):
        tulos = IntJoukko()

        for luku in a.to_int_list():
          if b.kuuluu(luku):
            tulos.lisaa(luku)

        return tulos

    @staticmethod
    def erotus(a, b):
        tulos = IntJoukko()

        for luku in a.to_int_list():
          if not b.kuuluu(luku):
            tulos.lisaa(luku)

        return tulos

    def __str__(self):
        if self.alkioiden_lkm == 0:
            return "{}"

        tuotos = "{"

        for i in range(0, self.alkioiden_lkm - 1):
            tuotos += f"{self.luvut[i]}, "

        tuotos += f"{self.luvut[self.alkioiden_lkm - 1]}"+"}"
        return tuotos
